spyder/widgets/tabs.py

```diff
@@ -337,3 +357,85 @@ def move_tab_from_another_tabwidget(self, tabwidget_from,
         # (see Issue 1094, Issue 1098)
         self.sig_move_tab.emit(tabwidget_from, str(id(self)), index_from,
                                index_to)
+
+
+class EditTabNamePopup(QLineEdit):
+    """Popup on top of the tab to edit its name"""
+
+    def __init__(self, parent):
+        QLineEdit.__init__(self, parent=None)
+
+        self.main = parent if parent is not None else self.parent()
+      
+        # TabBar will have the focus
+        self.setFocusProxy(self.main)
+
+        left_margin = self.main.contentsRect().left()
+        self.setTextMargins(left_margin, 0, 0, 0)
+
+        # Slot to handle tab name update
+        self.editingFinished.connect(self.edit_finished)
+
+        # Even filter to catch clicks and ESC key
+        self.installEventFilter(self)
+        
+        # Clean borders and no shadow to blend with tab
+        self.setWindowFlags(
+            Qt.Popup |
+            Qt.FramelessWindowHint |
+            Qt.NoDropShadowWindowHint
+        )
+        self.setFrame(False)
+
+        # Align with tab name
+        self.setTextMargins(9, 0, 0, 0)
+
+        # Track with tab is being edited
+        self.tab_index = None

```

Comment:
It would be good to have a similar structure to how other widgets work

```

# Instance variables

# Widgets

# Widget setup

# Layouts

# Signals and slots

```